### PR TITLE
fix(components): [tag] leave animations not working

### LINKED ISSUE #17255: [Component] [tag] el-tag组件再移除时没有动画
Bug Type: **`Component`**

## Environment

- Vue Version: `3.4.21`
- Element Plus Version: `2.7.3`
- Browser / OS: `Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36`
- Build Tool: `Vite`

### Related Component

- `el-tag`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGRpdiBjbGFzcz1cImZsZXggZ2FwLTJcIj5cbiAgICA8ZWwtdGFnXG4gICAgICB2LWZvcj1cInRhZyBpbiBkeW5hbWljVGFnc1wiXG4gICAgICA6a2V5PVwidGFnXCJcbiAgICAgIGNsb3NhYmxlXG4gICAgICA6ZGlzYWJsZS10cmFuc2l0aW9ucz1cImZhbHNlXCJcbiAgICAgIEBjbG9zZT1cImhhbmRsZUNsb3NlKHRhZylcIlxuICAgID5cbiAgICAgIHt7IHRhZyB9fVxuICAgIDwvZWwtdGFnPlxuICAgIDxlbC1pbnB1dFxuICAgICAgdi1pZj1cImlucHV0VmlzaWJsZVwiXG4gICAgICByZWY9XCJJbnB1dFJlZlwiXG4gICAgICB2LW1vZGVsPVwiaW5wdXRWYWx1ZVwiXG4gICAgICBjbGFzcz1cInctMjBcIlxuICAgICAgc2l6ZT1cInNtYWxsXCJcbiAgICAgIEBrZXl1cC5lbnRlcj1cImhhbmRsZUlucHV0Q29uZmlybVwiXG4gICAgICBAYmx1cj1cImhhbmRsZUlucHV0Q29uZmlybVwiXG4gICAgLz5cbiAgICA8ZWwtYnV0dG9uIHYtZWxzZSBjbGFzcz1cImJ1dHRvbi1uZXctdGFnXCIgc2l6ZT1cInNtYWxsXCIgQGNsaWNrPVwic2hvd0lucHV0XCI+XG4gICAgICArIE5ldyBUYWdcbiAgICA8L2VsLWJ1dHRvbj5cbiAgPC9kaXY+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgbmV4dFRpY2ssIHJlZiB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB7IEVsSW5wdXQgfSBmcm9tICdlbGVtZW50LXBsdXMnXG5cbmNvbnN0IGlucHV0VmFsdWUgPSByZWYoJycpXG5jb25zdCBkeW5hbWljVGFncyA9IHJlZihbJ1RhZyAxJywgJ1RhZyAyJywgJ1RhZyAzJ10pXG5jb25zdCBpbnB1dFZpc2libGUgPSByZWYoZmFsc2UpXG5jb25zdCBJbnB1dFJlZiA9IHJlZjxJbnN0YW5jZVR5cGU8dHlwZW9mIEVsSW5wdXQ+PigpXG5cbmNvbnN0IGhhbmRsZUNsb3NlID0gKHRhZzogc3RyaW5nKSA9PiB7XG4gIGR5bmFtaWNUYWdzLnZhbHVlLnNwbGljZShkeW5hbWljVGFncy52YWx1ZS5pbmRleE9mKHRhZyksIDEpXG59XG5cbmNvbnN0IHNob3dJbnB1dCA9ICgpID0+IHtcbiAgaW5wdXRWaXNpYmxlLnZhbHVlID0gdHJ1ZVxuICBuZXh0VGljaygoKSA9PiB7XG4gICAgSW5wdXRSZWYudmFsdWUhLmlucHV0IS5mb2N1cygpXG4gIH0pXG59XG5cbmNvbnN0IGhhbmRsZUlucHV0Q29uZmlybSA9ICgpID0+IHtcbiAgaWYgKGlucHV0VmFsdWUudmFsdWUpIHtcbiAgICBkeW5hbWljVGFncy52YWx1ZS5wdXNoKGlucHV0VmFsdWUudmFsdWUpXG4gIH1cbiAgaW5wdXRWaXNpYmxlLnZhbHVlID0gZmFsc2VcbiAgaW5wdXRWYWx1ZS52YWx1ZSA9ICcnXG59XG48L3NjcmlwdD5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZSxcbiAgICBcImFsbG93SnNcIjogdHJ1ZSxcbiAgICBcImNoZWNrSnNcIjogdHJ1ZVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn1cbiIsIl9vIjp7fX0=)

### Steps to reproduce

点tag的关闭按钮移除tag

### What is Expected?

移除tag应该有动画

### What is actually happening?

移除tag应该无动画

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by makedopamine**
I filed an [issue](https://github.com/vuejs/core/issues/11703) to Vue and was aware that the problem is actually caused by a breaking change in Vue3.  I think we can either fix #17255 with monkey-patch like this PR, or remove leave animations (possibly even entire animations for consistency) to follow Transition design.

### CODE REVIEW

File: packages/components/tag/src/tag.vue [hunk 1]
```diff
@@ -12,7 +12,12 @@
       <Close />
     </el-icon>
   </span>
-  <transition v-else :name="`${ns.namespace.value}-zoom-in-center`" appear>
+  <transition
+    v-else
+    :name="`${ns.namespace.value}-zoom-in-center`"
+    appear
+    @vnode-mounted="handleVNodeMounted"
```
Comment (btea) [inline packages/components/tag/src/tag.vue:19]:
```suggestion
    @vue:mounted="handleVNodeMounted"
```

image

**Review by btea — APPROVED** [review]
LGTM.

File: packages/components/tag/src/tag.vue [hunk 2]
```diff
@@ -66,4 +72,9 @@ const handleClose = (event: MouseEvent) => {
 const handleClick = (event: MouseEvent) => {
   emit('click', event)
 }
+
+const handleVNodeMounted = (vnode: VNode) => {
+  // @ts-ignore
+  vnode.component.subTree.component.bum = null
```
Comment (btea) [inline packages/components/tag/src/tag.vue:78]: This operation of modifying the internal properties of a component may be adjusted in the future. Do you think we should keep this operation? @tolking

Comment (tolking) [reply packages/components/tag/src/tag.vue:78]: The probability of changing `bum(onBeforeUnmount)` is very small. 

We can fix the issues through this PR before finding a more suitable solution

**Review by tolking — APPROVED** [review]

**Comment by dr-charnog** [comment]
@makedopamine, this PR likely breaks tests. After migrating from version 2.8.0 to 2.8.2, our tests show the following error:

image

Vitest 2.0.5

**Comment by makedopamine** [comment]
> @makedopamine, this PR likely breaks tests. After migrating from version 2.8.0 to 2.8.2, our tests show the following error:

Could you provide a minimal reproduction?

**Comment by dr-charnog** [comment]
@makedopamine I think the main issue arises from the structure of the component. When the `v-if` directive wraps `<ElTag>` and 'hides' it, this problem occurs.

```vue
<ElTooltip v-if="<condition>">
  <ElTag />
</ElToolip>
```

Repo: https://github.com/dr-charnog/el-plus-tag-bum

**Comment by makedopamine** [comment]
@dr-charnog [VTU](https://github.com/vuejs/test-utils/) stubs transition by default. However, the implementation of transition stub has certain limitation as mentioned in the [document](https://test-utils.vuejs.org/guide/advanced/transitions.html#Partial-support). 
Here is a solution for your reference:
```diff
      global: {
        plugins: [ElementUI],
+        stubs: {
+         transition: false
+        }
      }
```

**Pull request merged**